README.md
# flutter_animated_cards

A Fully customizable animated cards widget that has feature of 3D animation and much more. This widget works on both Android & iOS.

## Installation

Add following dependency in pubspec.yaml file. And add this import to your file.

```yaml
flutter_animated_cards: ^0.0.1

import 'package:flutter_animated_cards/flutter_animated_cards.dart';
```
## Quick Start

```python
// Add AnimatedCards to the widget tree

child: FlutterAnimatedCards(
          list: _cardList,
          model: _cardModel,
        ),                ​
```

## Card Model

```dart
class CardModel {
   final String title;                  /// required
   final String imagePath;            /// required  i.e "assets/images/image.png"
   final String description;        /// (Optional)
   final Color color;             /// (Optional) add this parameter if you want to have different color on each card
   final List<Color> colorList; /// (Optional) add this parameter if you want to apply different gradient on each card
}
```
#### `FlutterAnimatedCards` made through model class
If you add Color parameter in your model class and assign each list item a new different color then your widget look like this.

Cards with ModelColor       |  Cards with Model Gradient         
:-------------------------:|:-----------------------------:
<img height="380px" src="https://user-images.githubusercontent.com/36657067/123776333-a2300380-d8e8-11eb-801a-6c4cd179bea3.gif?raw=true">|<img height="380px" src="https://user-images.githubusercontent.com/36657067/124254739-cb4bd080-db42-11eb-910b-37c1776d2f5e.gif?raw=true">

```python
1-
   CardModel(
           title: 'Colors, Vibrant',
           description:'Gardening is how I relax. It\'s another form of creating and playing with colors.',
           color: Color(0xfff9d9e2),  // add color of your choice, add different colors in every list item
           imagePath: "assets/images/colors.png"),

2- 
   CardModel(
          title: 'Budapest, Hungary',
          description:'Meet the CardModel with rich history and indescribable culture',
          colorList: [Color(0xFF4CAF50),  // add gradient of your choice, add different gradient in every list item
            Color(0xFF00BCD4),
            Color(0xFF99ca3d)],
          imagePath: "assets/images/city_image.png"),

```


#### `FlutterAnimatedCards` without passing any color in model class
This is another way to give colors or gradient to your card. Now you need to add 'cardWithSingleColor' or 'cardWithGradientColors' as shown below

 Cards with Single Color    |   Cards with Single Gradient               
:-------------------------:|:-----------------------------:
<img height="380px" src="https://user-images.githubusercontent.com/36657067/123781193-621f4f80-d8ed-11eb-80fc-f60da0c90bec.jpeg?raw=true">|<img height="380px" src="https://user-images.githubusercontent.com/36657067/123785579-20dd6e80-d8f2-11eb-8690-f9cc7bd91ee9.jpeg?raw=true">

```python
1-
   child: FlutterAnimatedCards(
          list: _cardList,
          model: _cardModel,
          cardWithSingleColor: Color(0xfff9d9e2),  // you are not asked to add this parameter, but you can this if you want to display your cards with any other single color. By default cards will appear as it is shown in Screenshot 
        ),

2- 
   child: FlutterAnimatedCards(
          list: _cardList,
          model: _cardModel,
          cardWithGradientColors: [Color(0xffacb6e5),  // this parameter is needs to be added to add gradient, there is no default gradient available in lib
           Color(0xFFD5D2D2),
           Color(0xFFfbc7d4)],
        ),

```

<br>

## Custom Usage
There are several options that allow for more control:

|  Properties  |   Description   |
|--------------|-----------------|
| `model` | This is the required parameter. You need to pass a model class somehow similar to the one shown in example, you can add more arguments in your class but 'name', 'description', 'image' is necessary |
| `list` | This is also the required parameter. You need to pass a list of your model class, this list can not be null or empty|
| `cardHeight` | set card height according to your requirement. |
| `cardWidth` | set card width according to your requirement. |
| `isRotatingCards` | by default it's 'true', you can set it 'false' if you don't want to add 3D rotation in your cards. |
| `cardWithSingleColor` | if you want to give all your cards one single color, use this property. You can also give each card a different color through model class |
| `cardWithGradientColors` | it takes List<Color> as a parameter, this property allows you to apply the same gradient on all of your cards. You can also give each card a different gradient through model class |
| `moreText` | by default I have set it to 'LEARN MORE', but you can change it to any text according to your choice. |
| `onMoreBtnPressed` | this is a click event, currently it performs nothing you can use it according to your choice. |


## Developer
Faiza Farooqui

## License
[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.md
## [0.0.1]  - initial release. Wednesday, 30 June 2021.

* This is the initial release of the 3D Animated Cards package. This includes features such as
- A animated cards that responds to user gestures
- fully Customizing the look and feel of the cards
- click events to perform desirable task


test/flutter_animated_cards_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_animated_cards/flutter_animated_cards.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/src/styles.dart
import 'package:flutter/material.dart';

class Styles {
  static const double hzScreenPadding = 18;

  static final TextStyle baseTitle =
      TextStyle(fontSize: 11, fontFamily: 'DMSerifDisplay');
  static final TextStyle baseBody =
      TextStyle(fontSize: 11, fontFamily: 'OpenSans');

  static final TextStyle appHeader =
      baseTitle.copyWith(color: Color(0xFF0e0e0e), fontSize: 30, height: 1);

  static final TextStyle cardTitle =
      baseTitle.copyWith(height: 1, color: Color(0xFF1a1a1a), fontSize: 25);
  static final TextStyle cardSubtitle =
      baseBody.copyWith(color: Color(0xFF666666), height: 1.5, fontSize: 12);
  static final TextStyle cardAction = baseBody.copyWith(
      color: Color(0xFFa6998b),
      fontSize: 10,
      height: 1,
      fontWeight: FontWeight.w600,
      letterSpacing: 0.1);

  static final TextStyle hotelsTitleSection = baseBody.copyWith(
      color: Colors.black,
      fontSize: 14,
      fontWeight: FontWeight.bold,
      height: 2);
  static final TextStyle hotelTitle = baseBody.copyWith(
      color: Colors.black, fontSize: 14, fontWeight: FontWeight.w600);
  static final TextStyle hotelPrice =
      baseBody.copyWith(color: Color(0xff4d4d4d), fontSize: 13);
  static final TextStyle hotelScore =
      baseBody.copyWith(color: Color(0xff0e0e0e));
  static final TextStyle hotelData = baseBody.copyWith(color: Colors.grey[700]);
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'model/city_model.dart';

import 'package:flutter_animated_cards/flutter_animated_cards.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Animated 3D Cards',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      debugShowCheckedModeBanner: false,
      home: SimpleCardDemo(),
    );
  }
}

class SimpleCardDemo extends StatefulWidget {
  @override
  _SimpleCardDemoState createState() => _SimpleCardDemoState();
}

class _SimpleCardDemoState extends State<SimpleCardDemo> {
  List<CardModel> _cardList;
  CardModel _cardModel;

  @override
  void initState() {
    var data = CardData();
    _cardList = data.getCardList();
    _cardModel = _cardList[1];

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text("Animated 3D Cards"),
        ),
        body: Center(
          child: FlutterAnimatedCards(
            list: _cardList,
            model: _cardModel,
          ),
        )
        );
  }
}

class FullyCustomizedCards extends StatefulWidget {
  @override
  _FullyCustomizedCardsState createState() => _FullyCustomizedCardsState();
}

class _FullyCustomizedCardsState extends State<FullyCustomizedCards> {
  List<CardModel> _cardList;
  CardModel _cardModel;

  @override
  void initState() {
    var data = CardData();
    _cardList = data.getCardList();
    _cardModel = _cardList[1];

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text("Animated 3D Cards"),
        ),
        body: Center(
          child: FlutterAnimatedCards(
            /// required
            list: _cardList,

            /// required
            model: _cardModel,

            /// either use this one
            cardWithSingleColor: Color(0xffdee5cf),

            /// or this one
            cardWithGradientColors: [
              Color(0xfffff060),
              Color(0xFFD5D2D2),
              Color(0xFFfbc7d4)
            ],

            /// set it false if you don't want 3D rotation
            isRotatingCards: true,

            /// change it according to your design
            cardHeight: 200,

            /// change it according to your design
            cardWidth: 170,

            moreText: 'Set any text here',
            onMoreBtnPressed: () {
              /// write a code that takes you to the detail page of this card
            },
          ),
        )

        /// bottomSheet: SlideUpBar(),
        );
  }
}


lib/src/rotation_3d.dart
import 'dart:math';

import 'package:flutter/material.dart';

/// Takes a x,y or z rotation, in degrees, and rotates. Good for spins & 3d flip effects
class Rotation3d extends StatelessWidget {
  /// Degrees to rads constant
  static const double degrees2Radians = pi / 180;

  final Widget child;
  final double rotationX;
  final double rotationY;
  final double rotationZ;

  const Rotation3d(
      {Key key,
      this.child,
      this.rotationX = 0,
      this.rotationY = 0,
      this.rotationZ = 0})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Transform(
        alignment: FractionalOffset.center,
        transform: Matrix4.identity()
          ..setEntry(3, 2, 0.001)
          ..rotateX(rotationX * degrees2Radians)
          ..rotateY(rotationY * degrees2Radians)
          ..rotateZ(rotationZ * degrees2Radians),
        child: child);
  }
}


lib/flutter_animated_cards.dart
library flutter_animated_cards;

import 'package:flutter/material.dart';
import 'src/rotation_3d.dart';
import 'src/styles.dart';

class FlutterAnimatedCards extends StatefulWidget {
  /// width and height of card
  double cardWidth, cardHeight;

  /// a bool var- if it's set to 'true' cards will show its 3D rotation pattern else it will behave as 2D
  bool isRotatingCards = true;

  /// background [Color] of card
  final Color cardWithSingleColor;

  /// list of [Color] used for apply gradient of card
  final List<Color> cardWithGradientColors;

  /// we need model to display information (text and images) on cards
  var model;
  var currentCardData;

  /// list to show number of cards and render data
  final List<dynamic> list;

  /// card click event. You can perform your desire action by using this click event when user is clicking on card
  Function onMoreBtnPressed;

  /// by default this text is set to 'LEARN MORE' you can change it to your choice
  String moreText;

  FlutterAnimatedCards({
    Key key,
    this.cardWidth: 160,
    this.cardHeight: 200,
    isRotatingCards: true,
    this.cardWithSingleColor: const Color(0xffdaf3f7),
    this.cardWithGradientColors: const <Color>[],
    @required this.list: const <dynamic>[],
    @required this.model,
    this.moreText: "Learn More",
    this.onMoreBtnPressed,
  }) : super(key: key);

  @override
  FlutterAnimatedCardsState createState() => FlutterAnimatedCardsState();
}

class FlutterAnimatedCardsState extends State<FlutterAnimatedCards>
    with SingleTickerProviderStateMixin {
  final double _maxRotation = 20;

  PageController _pageController;

  double _normalizedOffset = 0;
  double _prevScrollX = 0;
  bool _isScrolling = false;

  AnimationController _tweenController;
  Tween<double> _tween;
  Animation<double> _tweenAnim;

  @override
  void initState() {
    int tweenTime = 1000;
    _tweenController = AnimationController(
        vsync: this, duration: Duration(milliseconds: tweenTime));

    ///Create Tween, which defines our begin + end values
    _tween = Tween<double>(begin: -1, end: 0);

    widget.currentCardData = widget.list[1];

    super.initState();
  }

  void onCardChangeListener(var model) {
    setState(() {
      this.widget.currentCardData = model;
    });
  }

  @override
  Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;
    widget.cardHeight = (size.height * .48).clamp(300.0, 400.0);
    widget.cardWidth = widget.cardHeight * .8;

    ///Calculate the viewPort fraction for this aspect ratio, since PageController does not accept pixel based size values
    _pageController = PageController(
        initialPage: 1, viewportFraction: widget.cardWidth / size.width);

    ///Create our main list
    Widget listContent = Container(
      ///Wrap list in a container to control height and padding
      height: widget.cardHeight,

      ///Use a ListView.builder, calls buildItemRenderer() lazily, whenever it need to display a listItem
      child: PageView.builder(
        ///Use bounce-style scroll physics, feels better with this demo
        physics: BouncingScrollPhysics(),
        controller: _pageController,
        itemCount: widget.list.length,
        scrollDirection: Axis.horizontal,
        itemBuilder: (context, i) => _buildItemRenderer(widget.list[i]),
      ),
    );

    /// Wrap our list content in a Listener to detect PointerUp events, and a NotificationListener to detect ScrollStart and ScrollUpdate
    /// We have to use both, because NotificationListener does not inform us when the user has lifted their finger.
    /// We can not use GestureDetector like we normally would, ListView suppresses it while scrolling.
    return Listener(
      onPointerUp: _handlePointerUp,
      child: NotificationListener(
        onNotification: _handleScrollNotifications,
        child: listContent,
      ),
    );
  }

  ///Create a renderer for each list item
  Widget _buildItemRenderer(var dummyModel) {
    return Container(
      ///Vertically pad all the non-selected items, to make them smaller. AnimatedPadding widget handles the animation.
      child: Rotation3d(
        rotationY:
            widget.isRotatingCards ? _normalizedOffset * _maxRotation : 0,

        ///Create the actual content renderer for our list
        child: TravelCardRenderer(
          ///Pass in the offset, renderer can update it's own view from there
          _normalizedOffset,
          cardWidth: widget.cardWidth,
          model: dummyModel,
          cardHeight: widget.cardHeight,
          cardWithSingleColor: widget.cardWithSingleColor,
          cardWithGradientColors: widget.cardWithGradientColors,
          onMoreBtnPressed: widget.onMoreBtnPressed,
          moreText: widget.moreText,
        ),
      ),
    );
  }

  ///Check the notifications bubbling up from the ListView, use them to update our currentOffset and isScrolling state
  bool _handleScrollNotifications(Notification notification) {
    ///Scroll Update, add to our current offset, but clamp to -1 and 1
    if (notification is ScrollUpdateNotification) {
      if (_isScrolling) {
        double dx = notification.metrics.pixels - _prevScrollX;
        double scrollFactor = .01;
        double newOffset = (_normalizedOffset + dx * scrollFactor);
        _setOffset(newOffset.clamp(-1.0, 1.0));
      }
      _prevScrollX = notification.metrics.pixels;

      ///Calculate the index closest to middle
      ///_focusedIndex = (_prevScrollX / (_itemWidth + _listItemPadding)).round();
      onCardChangeListener(widget.list
          .elementAt(_pageController.page.round() % widget.list.length));
    }

    ///Scroll Start
    else if (notification is ScrollStartNotification) {
      _isScrolling = true;
      _prevScrollX = notification.metrics.pixels;
      if (_tween != null) {
        _tweenController.stop();
      }
    }
    return true;
  }

  ///If the user has released a pointer, and is currently scrolling, we'll assume they're done scrolling and tween our offset to zero.
  ///This is a bit of a hack, we can't be sure this event actually came from the same finger that was scrolling, but should work most of the time.
  void _handlePointerUp(PointerUpEvent event) {
    if (_isScrolling) {
      _isScrolling = false;
      _startOffsetTweenToZero();
    }
  }

  ///Helper function, any time we change the offset, we want to rebuild the widget tree, so all the renderers get the new value.
  void _setOffset(double value) {
    setState(() {
      _normalizedOffset = value;
    });
  }

  ///Tweens our offset from the current value, to 0
  void _startOffsetTweenToZero() {
    ///The first time this runs, setup our controller, tween and animation. All 3 are required to control an active animation.
    int tweenTime = 1000;
    if (_tweenController == null) {
      ///Create Controller, which starts/stops the tween, and rebuilds this widget while it's running
      _tweenController = AnimationController(
          vsync: this, duration: Duration(milliseconds: tweenTime));

      ///Create Tween, which defines our begin + end values
      _tween = Tween<double>(begin: -1, end: 0);

      ///Create Animation, which allows us to access the current tween value and the onUpdate() callback.
      _tweenAnim = _tween.animate(new CurvedAnimation(
          parent: _tweenController, curve: Curves.elasticOut))

        ///Set our offset each time the tween fires, triggering a rebuild
        ..addListener(() {
          _setOffset(_tweenAnim.value);
        });
    }

    ///Restart the tweenController and inject a new start value into the tween
    _tween.begin = _normalizedOffset;
    _tweenController.reset();
    _tween.end = 0;
    _tweenController.forward();
  }
}

class TravelCardRenderer extends StatelessWidget {
  final double offset;
  final double cardWidth;
  final double cardHeight;
  final Color cardWithSingleColor;
  final List<Color> cardWithGradientColors;
  var model;
  Function onMoreBtnPressed;
  String moreText;

  TravelCardRenderer(
    this.offset, {
    Key key,
    this.cardWidth = 250,
    this.cardHeight,
    this.cardWithSingleColor,
    this.cardWithGradientColors: const <Color>[],
    this.model,
    this.moreText,
    this.onMoreBtnPressed,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    print(
        '======================= cardWithGradientColors  ${cardWithGradientColors}');
    print('======================= model.colorList  ${model.colorList}');
    LinearGradient _linearGradient;
    if (model.colorList != null && model.colorList.length > 0) {
      _linearGradient = LinearGradient(
          colors: model.colorList,
          stops: (model.colorList.length == 0)
              ? []
              : (model.colorList.length == 1)
                  ? [1.0]
                  : (model.colorList.length == 2)
                      ? [0.5, 1.0]
                      : [0.0, 0.5, 1.0],
          tileMode: TileMode.mirror,
          end: Alignment.bottomCenter,
          begin: Alignment.topRight);
    } else if (cardWithGradientColors != null &&
        cardWithGradientColors.length > 0) {
      _linearGradient = LinearGradient(
          colors: cardWithGradientColors,
          stops: (cardWithGradientColors.length == 0)
              ? []
              : (cardWithGradientColors.length == 1)
                  ? [1.0]
                  : (cardWithGradientColors.length == 2)
                      ? [0.5, 1.0]
                      : [0.0, 0.5, 1.0],
          tileMode: TileMode.mirror,
          end: Alignment.bottomCenter,
          begin: Alignment.topRight);
    } else {
      _linearGradient = null;
    }
    return Container(
      width: cardWidth,
      margin: EdgeInsets.only(top: 8),
      child: Stack(
        overflow: Overflow.visible,
        alignment: Alignment.center,
        children: <Widget>[
          // Card background color & decoration
          Container(
            margin: EdgeInsets.only(top: 30, left: 12, right: 12, bottom: 12),
            decoration: BoxDecoration(
              gradient: _linearGradient,
              color: model.color == null ? cardWithSingleColor : model.color,
              borderRadius: BorderRadius.circular(8),
              boxShadow: [
                BoxShadow(color: Colors.black12, blurRadius: 4 * offset.abs()),
                BoxShadow(
                    color: Colors.black12, blurRadius: 10 + 6 * offset.abs()),
              ],
            ),
          ),
          // City image, out of card by 15px
          Positioned(top: -15, child: _buildCityImage()),
          // City information
          _buildCityData()
        ],
      ),
    );
  }

  Widget _buildCityImage() {
    double maxParallax = 30;
    double globalOffset = offset * maxParallax * 2;
    double cardPadding = 28;
    double containerWidth = cardWidth - cardPadding;
    return Container(
      height: cardHeight,
      width: containerWidth,
      child: Stack(
        alignment: Alignment.bottomCenter,
        children: <Widget>[
          // _buildPositionedLayer("assets/images/${city.name}/${city.name}-Back.png", containerWidth * .8, maxParallax * .1, globalOffset),
          // _buildPositionedLayer("assets/images/${city.name}/${city.name}-Middle.png", containerWidth * .9, maxParallax * .6, globalOffset),
          // _buildPositionedLayer("assets/images/${city.name}/${city.name}-Front.png", containerWidth * .9, maxParallax, globalOffset),
          _buildPositionedLayer(
              model.imagePath, containerWidth * .9, maxParallax, globalOffset),
        ],
      ),
    );
  }

  Widget _buildCityData() {
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      children: <Widget>[
        // The sized box mock the space of the city image
        SizedBox(width: double.infinity, height: cardHeight * .57),
        Padding(
          padding: const EdgeInsets.symmetric(horizontal: 12.0),
          child: Text(
            model.title,
            style: Styles.cardTitle,
            textAlign: TextAlign.center,
            maxLines: 1,
          ),
        ),
        Expanded(
          child: Padding(
            padding: const EdgeInsets.symmetric(horizontal: 25.0),
            child: Text(
              model.description,
              style: Styles.cardSubtitle,
              textAlign: TextAlign.center,
              maxLines: 3,
            ),
          ),
        ),
        // Expanded(child: SizedBox(),),
        FlatButton(
          disabledColor: Colors.transparent,
          color: Colors.transparent,
          child: Text(moreText.toUpperCase(), style: Styles.cardAction),
          onPressed: () {
            if (onMoreBtnPressed != null) {
              onMoreBtnPressed();
            }
          },
        ),
        SizedBox(height: 8)
      ],
    );
  }

  Widget _buildPositionedLayer(
      String path, double width, double maxOffset, double globalOffset) {
    double cardPadding = 24;
    double layerWidth = cardWidth - cardPadding;
    return Positioned(
        left: ((layerWidth * .5) - (width / 2) - offset * maxOffset) +
            globalOffset,
        bottom: cardHeight * .45,
        child: Image.asset(
          path,
          width: width,
          // package: App.pkg,
        ));
  }
}


example/lib/model/city_model.dart
import 'package:flutter/material.dart';

class CardModel {
  final String title;

  /// required
  final String imagePath;

  /// required
  final String description;

  /// (Optional)
  final Color color;

  /// (Optional) add this parameter if you want to have different color on each card
  final List<Color> colorList;

  /// (Optional) add this parameter if you want to apply different gradient on each card

  CardModel(
      {this.title,
      this.description: "",
      this.color,
      this.colorList: const <Color>[],
      this.imagePath});
}

class CardData {
  List<CardModel> _cardData = [
    CardModel(
        title: 'Colors, Vibrant',
        description:
            'Gardening is how I relax. It\'s another form of creating and playing with colors.',
        // color: Color(0xfff9d9e2), // 0xffdee5cf
        colorList: [Color(0xffacb6e5), Color(0xFFD5D2D2), Color(0xFFfbc7d4)],
        imagePath: "assets/images/colors.png"),
    CardModel(
        title: 'Budapest, Hungary',
        description:
            'Meet the CardModel with rich history and indescribable culture',
        colorList: [Color(0xFF4CAF50), Color(0xFF00BCD4), Color(0xFF99ca3d)],
        imagePath: "assets/images/city_image.png"),
    CardModel(
        title: 'London, England',
        description:
            'A diverse and exciting CardModel with the world’s best sights and attractions!',
        // color: Color(0xffdee5cf),
        colorList: [Color(0xfffff060), Color(0xFFD5D2D2), Color(0xFFfbc7d4)],
        imagePath: "assets/images/london_city.png"),
  ];

  List<CardModel> getCardList() => _cardData;
}
